e2e/testcases/basic_test.go

```diff
@@ -0,0 +1,318 @@
+		nt.T.Fatal(err)
+	}
+}
+
+func TestNamespaceGarbageCollection(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation2)
+
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/accounting-namespace.yaml", yamlDir), "acme/namespaces/accounting/namespace.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Add accounting namespace")
+	nt.WaitForRepoSyncs()
+
+	if err := nt.Validate("accounting", "", &corev1.Namespace{}); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.RootRepos[configsync.RootSyncName].Remove("acme/namespaces/accounting/namespace.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Remove accounting namespace")
+	nt.WaitForRepoSyncs()
+
+	if err := nt.ValidateNotFound("accounting", "", &corev1.Namespace{}); err != nil {
+		nt.T.Fatal("Namespace still exist when it should have been garbage collected")
+	}
+}
+
+func TestNamespacePolicyspaceConversion(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation2)
+
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/dir-namespace.yaml", yamlDir), "acme/namespaces/dir/namespace.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Add dir namespace")
+	nt.WaitForRepoSyncs()
+
+	if err := nt.Validate("dir", "", &corev1.Namespace{}); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	nt.RootRepos[configsync.RootSyncName].Remove("acme/namespaces/dir/namespace.yaml")
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/subdir-namespace.yaml", yamlDir), "acme/namespaces/dir/subdir/namespace.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Remove dir namespace, add subdir namespace")
+	nt.WaitForRepoSyncs()
+
+	if err := nt.Validate("subdir", "", &corev1.Namespace{}); err != nil {
+		nt.T.Fatal(err)
+	}
+
+	if err := nt.ValidateNotFound("dir", "", &corev1.Namespace{}); err != nil {
+		nt.T.Fatal("Namespace still exist when it should have been converted")
+	}
+}
+
+func TestSyncDeploymentAndReplicaSet(t *testing.T) {
+	nt := nomostest.New(t, nomostesting.Reconciliation2)
+
+	// Test the ability to fix a mistake: overlapping replicaset and deployment.
+	// Readiness behavior is undefined for this race condition.
+	// One or both of the Deployment and ReplicaSet may become unhealthy.
+	// But regardless, the user should be able to correct the situation.
+	nt.T.Log("Add a replicaset")
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/dir-namespace.yaml", yamlDir), "acme/namespaces/dir/namespace.yaml")
+	nt.RootRepos[configsync.RootSyncName].Copy(fmt.Sprintf("%s/replicaset-helloworld.yaml", yamlDir), "acme/namespaces/dir/replicaset.yaml")
+	nt.RootRepos[configsync.RootSyncName].CommitAndPush("Add replicaset")
+
+	// This sync may block until reconcile timeout is reached,
+	// because ReplicaSet or Deployment may never reconcile.
+	// So this wait timeout must be longer than the reconcile timeout (5m).
+	nt.WaitForRepoSyncs()
```

Comment:
Does this not need to have a longer timeout like the comment suggests?

Reply:
it looks like the `DefaultWaitTimeout` for `WaitForRepoSyncs` is 6 minutes, so i think we should be fine. Let me know if you think this may be a concern

Reply:
That should be fine